packages/opentelemetry-types/src/trace/span.ts

```diff
@@ -0,0 +1,106 @@
+/**
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { SpanContext } from './span_context';
+import { Status } from './status';
+
+/**
+ * An interface that represents a span. A span represents a single operation
+ * within a trace. Examples of span might include remote procedure calls or a
+ * in-process function calls to sub-components. A Trace has a single, top-level
+ * "root" Span that in turn may have zero or more child Spans, which in turn
+ * may have children.
+ */
+export interface Span {
+  /**
+   * Returns the SpanContext object associated with this Span.
+   *
+   * @returns the SpanContext object associated with this Span.
+   */
+  context(): SpanContext;
+
+  // /**
+  //  * # TODO
+  //  * Returns the Tracer object used to create this Span.
+  //  * https://github.com/open-telemetry/opentelemetry-specification/issues/21
+  //  */
+  // tracer(): Tracer;
+
+  /**
+   * Sets an attribute to the span.
+   *
+   * @param key the key for this attribute.
+   * @param value the value for this attribute. If the value is a typeof object
+   *     it has to be JSON.stringify-able otherwise it will raise an exception.
+   */
+  setAttribute(key: string, value: string | number | boolean | object): void;
+
+  /**
+   * Adds an event to the Span.
+   *
+   * @param name the name of the event.
+   * @param [attributes] the attributes that will be added; these are
+   *     associated with this event.
+   */
+  addEvent(
+    name: string,
+    attributes?: { [key: string]: string | number | boolean | object }
+  ): void;
+
+  /**
+   * Adds a link to the Span.
+   *
+   * @param spanContext the context of the linked span.
+   * @param [attributes] the attributes that will be added; these are
+   *     associated with this link.
+   */
+  addLink(
+    spanContext: SpanContext,
```

Comment:
This should accept a `Span` as well.

Reply:
How is the type of link determined?

Reply:
> How is the type of link determined?

Good Point! AFAIK will use `attributes` to highlight the link type and will handle in `sdk` implementation. @songy23 and @rghetia to add more on this.

> This should accept a Span as well.

Can you please mention the use case with `span` ?

Reply:
> Can you please mention the use case with span ?

In general, I think everywhere a `SpanContext` is accepted should also accept a `Span`. This avoids the caller having to always call `span.context()` if they are referencing a span. I don't have a specific use case but I've had a lot of cases where I was happy that OpenTracing did that for me. This is especially true if the context propagator returns the active span.